### DISCUSSION TITLE
Issues with reading PixelData

### DISCUSSION THREAD

**Discussion by lenniuhr** [fo-dicom/fo-dicom#1629] [Q&A] [answered]
**Describe the issue**
My goal is to access image data of a DICOM with 16 bit depth. After researching I found out, that I need to work directly with the PixelData to achieve this. I tried this and faced some issues:

```C#
DicomFile dicomFile = DicomFile.Open(path);
DicomPixelData dicomPixelData = DicomPixelData.Create(dicomFile.Dataset);
IPixelData iPixelData = PixelDataFactory.Create(dicomPixelData, 0);

for (int x = 0; x < iPixelData.Width; x++)
{
    for (int y = 0; y < iPixelData.Height; y++)
    {
        double value = iPixelData.GetPixel(x, y);
        //...
    }
}
```

1. For some data, PixelDataFactory.Create had this error: "IndexOutOfRangeException: Index was outside the bounds of the array."
2. For other data, the method GetPixel had the error "IndexOutOfRangeException: Index was outside the bounds of the array.", although I'm staying within the Width and Height of the iPixelData object. I found out that for my 512x512 image it was still possible to access GetPixel(126382, 0), which ist very strange and seems to be connected to the raw size of 252766 bytes (it's exactly the half).

**Expected behavior**
I would expect, that it's possible to read a value for each pixel in the PixelData, corresponding to the image size.

**Screenshots or test DICOM files**
This zip contains 2 DICOM files, 1 for each issue.
[dicoms.zip](https://github.com/fo-dicom/fo-dicom/files/12064055/dicoms.zip)

**Environment**
I am using fo-dicom 5.1.1 in Unity 2021.3 on Windows 11.

**Comment by mrbean-bremen** [comment, marked as answer]
Your datasets are compressed, and if you directly access `PixelData` as you do, you get the compressed data (which is naturally smaller than the uncompressed data, and not what you need).
You can instead use `ImageData` to access the needed data:
```cs
var imageData = new ImageData(path);
IPixelData iPixelData = imageData.Pixels;

for (int x = 0; x < iPixelData.Width; x++)
...
```
which does the decompression on the fly if needed.

**Reply by lenniuhr** [reply]
Thanks so much! ImageData is exactly what I was looking for, I must have missed it.
